File: 76-minimum-window-substring/minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        m,n=len(s),len(t)
        if n>m:
            return ""
        freq=[0]*128
        for ch in t:
            freq[ord(ch)]+=1
        left=0
        count=0
        min_len=float('inf')
        start_index=-1
        for right in range(m):
            ch=s[right]
            if freq[ord(ch)]>0:
                count+=1
            freq[ord(ch)]-=1
            while count==n:
                window_len=right-left+1
                if window_len<min_len:
                    min_len=window_len
                    start_index=left
                left_char=s[left]
                freq[ord(left_char)]+=1
                if freq[ord(left_char)]>0:
                    count-=1
                left+=1
        return "" if start_index==-1 else s[start_index:start_index+min_len]

File: 76-minimum-window-substring/minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {

        int m = s.length();
        int n = t.length();

        if (n > m) return "";

        // Frequency array for ASCII characters
        int[] freq = new int[128];

        // Count characters in t
        for (char ch : t.toCharArray()) {
            freq[ch]++;
        }

        int left = 0;
        int count = 0;

        int minLen = Integer.MAX_VALUE;
        int startIndex = -1;

        // Expand window
        for (int right = 0; right < m; right++) {

            char ch = s.charAt(right);

            // Needed character found
            if (freq[ch] > 0) {
                count++;
            }

            freq[ch]--;

            // Valid window found
            while (count == n) {

                // Update minimum answer
                int windowLen = right - left + 1;

                if (windowLen < minLen) {
                    minLen = windowLen;
                    startIndex = left;
                }

                char leftChar = s.charAt(left);

                // Remove left character
                freq[leftChar]++;

                // Required character removed
                if (freq[leftChar] > 0) {
                    count--;
                }

                left++;
            }
        }

        return startIndex == -1
                ? ""
                : s.substring(startIndex, startIndex + minLen);
    }
}